Fix lua errors in serializer

ligurio opened the pull request.

Buristan commented on test/fuzz/luaL_loadbuffer/serializer.cc:
```diff
@@ -113,6 +115,80 @@ PROTO_TOSTRING(UnaryOperator, op);
+		retval += methods[i];
+		retval += "=";
+		retval += "function()\n";
+		retval += "local retval = {};\n";
```
Nit: `;` is excess since `\n` is used.
Here and below.

Buristan commented on test/fuzz/luaL_loadbuffer/serializer.cc:
```diff
@@ -113,6 +115,80 @@ PROTO_TOSTRING(UnaryOperator, op);
+		retval += ",\n";
+	}
+
+	retval += "}\n";
```
I suggest to use something like the following structure:
```lua
local metatable = {}
local function metamethod()
        return debug.setmetatable({}, metatable)
end 
metatable.__add = metamethod
-- ...
metatable.__unm = metamethod
```

It becomes much more easier to read. (AFAIKS you use the same approach in the last commit.)

ligurio replied:
Fixed in `test/fuzz/luaL_loadbuffer/mt.lua`.

Buristan commented on test/fuzz/luaL_loadbuffer/serializer.cc:
```diff
@@ -24,6 +24,8 @@ namespace luajit_fuzzer {
 namespace {
 
 const std::string kCounterNamePrefix = "counter_";
+const std::string kWrapperName = "wrap_val";
```
Nit: I prefer to name this as `wrap_with_table` (or something like that) for better understanding. 
Feel free to ignore.

Also, comments about these wrappers are desirable.

ligurio replied:
Lua function `wrap_val` was rewritten, now it adds a number of metamethods and return a table when passed value is `nil`. I'm not sure that `wrap_with_table` is a suitable name now. However, `wrap_val` is not a good choice too.

Left as is.

Buristan commented on test/fuzz/luaL_loadbuffer/serializer.cc:
```diff
@@ -113,6 +115,80 @@ PROTO_TOSTRING(UnaryOperator, op);
+	retval += "}\n";
+
+	retval += kWrapperName;
+	retval += " = function(val) local retval = {val}; ";
```
Maybe it is better to set metatable on raw values for tables?
So, we don't need an additional abstraction level for them.
```lua
local retval = val
if type(val) ~= 'table' then
  retval = {val}
end
```

ligurio replied:
this part of code is gone

Buristan reviewed (changes requested):
Hi, Sergey!
Thanks for the patchset!
It is nice to see that so many errors are fixed without dramatic efforts!

I'll proceed with the review per patch below.

**[PATCH 1/5] test/fuzz: fix a group of semantic errors in serializer**

I like the approach itself, just have a few stylistic suggestions.

Buristan commented on test/fuzz/luaL_loadbuffer/serializer.cc:
```diff
@@ -647,6 +737,26 @@ PROTO_TOSTRING(ForCycleName, forcyclename)
+			explist.explast());
+	} else {
+		explist_str += WrappedExpresionToString(
+			explist.expressions(0)) + "), ";
```
Typo: looks like extra `)` here.

ligurio replied:
Squash with a previous commit.

Buristan reviewed (changes requested):
**[PATCH 2/5]  test/fuzz: fix "'do' expected near ')'" in serializer**
The `ForCycleListExpressionsToString()` was introduced in the previous commit, so this commit should be squashed with it.

Buristan reviewed (changes requested):
**[PATCH 3/5]  test/fuzz: fix "cannot use '...' outside a vararg function near '...'"**

The commit message is more than 50 symbols.

Also, I don't like the way the issue is resolved -- it restrics the usage of any VARARG functions in the fuzzer.
I suppose it is better to add an additional flag to the block object, like it was done for the `return` and `break` issue, so this case will be ignored only for non-vararg functions.

Buristan reviewed (approved):
**[PATCH 4/5]  test/fuzz: fix "table index is nil" in serializer**
LGTM.

Buristan commented on test/fuzz/luaL_loadbuffer/mt.lua:
```diff
@@ -0,0 +1,38 @@
+mt.__unm = mt_method
+
+local wrap_val = function(val)
+    debug.setmetatable(val, mt);
```
I see a little bit changes since the 1st commit.
Still some errors can be observed like `"compare nil with string"`, since `val` may be `nil` value.
If use the following approach I have much less errors (0 for 20000 runs):
```lua
local wrap_val = function(val)
    if type(val) ~= table then
        val = {val}
    end
    debug.setmetatable(val, mt);
    return val
end
```

ligurio replied:
I believe these errors should gone with commit:

```
commit bbe1a266d2c8ff6b1aa1523ecf06882b9f92b6a1
Author: Sergey Bronnikov <sergeyb@tarantool.org>
Date:   Tue Sep 26 14:19:45 2023 +0300

    test/fuzz: fix a group of semantic errors in serializer
```

Buristan replied:
So, this `wrap_val()` function is not very usefull. May we try to remove it at all?

Buristan reviewed (changes requested):
**[PATCH 5/5] test/fuzz: lua prerequisites in external file [TO SQUASH]**

I like the idea of using simple Lua file to include -- it's much easier to read and change.

But I see some difference between the first version of the metamethod function and this one, see one comment below.

Buristan commented:
Also, I suppose we should fix all issues found by fuzzer first, because this patchset differs code semantics, so some protobuf files with reproducers are no longer useful.

Thoughts?

ligurio commented on test/fuzz/luaL_loadbuffer/mt.lua:
```diff
@@ -0,0 +1,38 @@
+local mt = {}
+
+local mt_method = function()
+   local retval = {};
```
removed `;` everywhere

ligurio commented:
> 

> [PATCH 3/5] test/fuzz: fix "cannot use '...' outside a vararg function near '...'"

> The commit message is more than 50 symbols.

I don't know how to fix it, commit header contains an error message. I don't want to rephrase with my own wording.

> Also, I don't like the way the issue is resolved -- it restrics the usage of any VARARG functions in the fuzzer. I suppose it is better to add an additional flag to the block object, like it was done for the return and break issue, so this case will be ignored only for non-vararg functions.

Fixed in another way (the same way as Rimma fixed issue with `break`, see 8df690b8824a0329cecadc39743d3b3577f85c33).

ligurio commented on test/fuzz/luaL_loadbuffer/preamble.lua:
```diff
@@ -0,0 +1,193 @@
+
+local wrap_binop = function(v1, v2, binop)
+    local is_equal_type = type(v1) == type(v2)
+    if is_equal_type == false then
```
Updated:

```lua
local wrap_binop = function(v1, v2, binop)
    if type(v1) ~= type(v2) then
        return false
    end
    if type(v1) == 'number' then
        return loadstring(('return %s %s %s'):format(v1, binop, v2))()
    end
    return false
end
```

ligurio commented:
"checkpatch always failed due to a missed/added spaces around &"
https://github.com/tarantool/checkpatch/issues/68

Buristan reviewed (changes requested):
Hi, Sergey!
Thanks for the fixes!
Please consider my comments below.

ligurio commented on test/fuzz/luaL_loadbuffer/preamble.lua:
```diff
@@ -0,0 +1,172 @@
+local def_number = 1
```
Fixed:

```diff
diff --git a/test/fuzz/luaL_loadbuffer/preamble.lua b/test/fuzz/luaL_loadbuffer/preamble.lua
index 2e6bfcde82..77a86bea5f 100644
--- a/test/fuzz/luaL_loadbuffer/preamble.lua
+++ b/test/fuzz/luaL_loadbuffer/preamble.lua
@@ -1,7 +1,7 @@
-local def_number = 1
+local DEFAULT_NUMBER = 1
 
 local wrap_number = function(val)
-    return type(val) == 'number' and val or def_number
+    return type(val) == 'number' and val or DEFAULT_NUMBER
 end
 
 local __add = function(v1, v2)
@@ -14,7 +14,7 @@ local __concat = function(v1, v2)
     return tostring(v1) .. tostring(v2)
 end
 local __div = function(v1, v2)
-    v2 = v2 == 0 and def_number or v2
+    v2 = v2 == 0 and DEFAULT_NUMBER or v2
     return wrap_number(v1) / wrap_number(v2)
 end
 local __eq = function(v1, v2)
@@ -36,7 +36,7 @@ local __le = function(v1, v2)
     end
 end
 local __len = function(v)
-    return type(v) == 'string' and #v or def_number
+    return type(v) == 'string' and #v or DEFAULT_NUMBER
 end
 local __lt = function(v1, v2)
     if type(v1) == 'number' and type(v2) == 'number' then
@@ -48,7 +48,7 @@ local __lt = function(v1, v2)
     end
 end
 local __mod = function(v1, v2)
-    v2 = v2 == 0 and def_number or v2
+    v2 = v2 == 0 and DEFAULT_NUMBER or v2
     return wrap_number(v1) % wrap_number(v2)
 end
 local __mul = function(v1, v2)
```

```diff
--- a/test/fuzz/luaL_loadbuffer/preamble.lua
+++ b/test/fuzz/luaL_loadbuffer/preamble.lua
@@ -5,7 +5,7 @@ local wrap_number = function(val)
 end
 
 local wrap_nan_nil = function(val)
-    return (val ~= val or val == nil) and def_number or val
+    return (val ~= val or val == nil) and DEFAULT_NUMBER or val
 end
 
 local __add = function(v1, v2)
```

ligurio commented on .luacheckrc:
```diff
@@ -108,3 +108,9 @@ files["test/replication/replica_quorum.lua"] = {
     },
 }
+
+files["test/fuzz/luaL_loadbuffer/preamble.lua"] = {
```
Added:

```diff
--- a/.luacheckrc
+++ b/.luacheckrc
@@ -109,6 +109,8 @@ files["test/replication/replica_quorum.lua"] = {
     },
 }
 
+-- Defined functions is unused in preamble,
+-- but could be used by generated Lua programs.
 files["test/fuzz/luaL_loadbuffer/preamble.lua"] = {
     ignore = {
         "211",
```

ligurio commented on test/fuzz/luaL_loadbuffer/preamble.lua:
```diff
@@ -0,0 +1,172 @@
+local def_number = 1
+
+local wrap_number = function(val)
+    return type(val) == 'number' and val or def_number
```
Replaced.

ligurio commented on test/fuzz/luaL_loadbuffer/preamble.lua:
```diff
@@ -0,0 +1,172 @@
+  __call = __call,
+  __div = __div,
+  __index = __index,
+  __len = __len,
```
Updated:

```diff
@@ -36,7 +36,7 @@ local __le = function(v1, v2)
     end
 end
 local __len = function(v)
-    return type(v) == 'string' and #v or DEFAULT_NUMBER
+    return DEFAULT_NUMBER
 end
 local __lt = function(v1, v2)
     if type(v1) == 'number' and type(v2) == 'number' then
```

and `__len` is removed for strings.

ligurio commented on test/fuzz/luaL_loadbuffer/preamble.lua:
```diff
@@ -0,0 +1,172 @@
+    return tostring(v1) .. tostring(v2)
+end
+local __div = function(v1, v2)
+    v2 = v2 == 0 and def_number or v2
```
Removed:

```diff
 local __div = function(v1, v2)
-    v2 = v2 == 0 and DEFAULT_NUMBER or v2
     return wrap_number(v1) / wrap_number(v2)
 end

 local __mod = function(v1, v2)
-    v2 = v2 == 0 and DEFAULT_NUMBER or v2
     return wrap_number(v1) % wrap_number(v2)
 end
```

ligurio commented on test/fuzz/luaL_loadbuffer/preamble.lua:
```diff
@@ -0,0 +1,172 @@
+end
+local __lt = function(v1, v2)
+    if type(v1) == 'number' and type(v2) == 'number' then
+        return v1 <= v2 -- numeric comparison
```
Done:

```diff
 local __lt = function(v1, v2)
     if type(v1) == 'number' and type(v2) == 'number' then
-        return v1 <= v2 -- numeric comparison
+        return v1 < v2 -- numeric comparison
     elseif type(v1) == 'string' and type(v2) == 'string' then
-        return v1 <= v2 -- lexicographic comparison
+        return v1 < v2 -- lexicographic comparison
     else
         return wrap_number(v1) <= wrap_number(v2)
     end
 end
```

ligurio commented on test/fuzz/luaL_loadbuffer/preamble.lua:
```diff
@@ -0,0 +1,172 @@
+end
+local __le = function(v1, v2)
+    if type(v1) == 'number' and type(v2) == 'number' then
+        return v1 <= v2 -- numeric comparison
```
Fixed.

ligurio commented on test/fuzz/luaL_loadbuffer/preamble.lua:
```diff
@@ -0,0 +1,172 @@
+    v2 = v2 == 0 and def_number or v2
+    return wrap_number(v1) / wrap_number(v2)
+end
+local __eq = function(v1, v2)
```
Now `__eq` returns `false` always.

Buristan replied:
So, we can drop it, since it returns the same value (`false`) for objects of different types.

ligurio replied:
Right, remove function `__eq`.

ligurio commented on test/fuzz/luaL_loadbuffer/preamble.lua:
```diff
@@ -0,0 +1,172 @@
+  __unm = __unm,
+})
+
+local function wrap_val(val)
```
`wrap_val` is gone, replaced with table `table_mt` in `preamble.lua` and setting this mt in serializer.

Buristan replied:
Nice!
Please, drop it also in the comment inside `NESTED_PROTO_TOSTRING(NameAssignment, assignment, Field)`.

ligurio replied:
Already removed.

Buristan commented on test/fuzz/luaL_loadbuffer/serializer.cc:
```diff
@@ -1011,7 +1045,14 @@ NESTED_PROTO_TOSTRING(ExpressionAssignment, assignment, Field)
-	assignment_str += " = " + ExpressionToString(assignment.value());
+	/*
+	 * There are cases when the same name is used for a value in
+	 * a table and then for a method: `wrap_val({a = {}}):a()`.
```
Since `wrap_val` is gone, this comment should be updated.

ligurio replied:
Removed.

Buristan reviewed (approved):
**[PATCH 1/6] Add @ligurio as codeowner for fuzzing tests**
Hi, Sergey!
Thanks for the patch!
LGTM!

Buristan reviewed (commented):
**[PATCH  2/6] test/fuzz: fix "cannot use '...' outside a vararg function"**

LGTM, with a minor comment below.

Buristan reviewed (approved):
**[PATCH 4/6] test/fuzz: fix "attempt to compare" in serializer**

Thanks for the patch!
LGTM with a minor nits regarding the commit message.

> There is a number of Lua errors related to comparison of non-comparable

Typo? s/There is/There are/
Typo: s/comparison/the comparison/

>  In Lua 5.1 and LuaJIT 2.1, that follows Lua 5.1, it is not possible
to overload Lua events `__le` and `__lt`, therefore an approach with
metamethods doesn't help there.

I suggest rewording like the following:

In Lua 5.1 and LuaJIT 2.1 (which follow the semantics and behaviour of Lua 5.1), it is not possible to overload Lua metamethods `__le` and `__lt` for these base types. Therefore, an approach with metamethods doesn't help there.

Buristan reviewed (approved):
**[PATCH 5/6] test/fuzz: fix "'<name>' expected near 'and'" in serializer**

Thanks for the patch!
LGTM, with several minor nits regarding the commit message.

>to an errors like "'<name>'expected near 'and'" because reserved

Typo: s/an//

>t.or = 0  -- ''<name>'' expected near ''or''

I suggest using the same format as in the commit title and in the sentence above -- it is more readable.

>  The patch fixes serializer, if automatically generated index name is

Typo: s/automatically/ the automatically/

equal to reserved keyword then postfix "_1" is added to index name.

Typo: s/reserved/the reserved/

ligurio commented on test/fuzz/luaL_loadbuffer/preamble.lua:
```diff
@@ -4,10 +4,6 @@ local always_number = function(val)
     return tonumber(val) or DEFAULT_NUMBER
 end
 
-local not_nan_and_nil = function(val)
```
Fixed, thanks. It is an accidental change.

ligurio commented on test/fuzz/luaL_loadbuffer/preamble.lua:
```diff
@@ -0,0 +1,166 @@
+    return always_number(v1) - always_number(v2)
+end
+local __unm = function(v)
+    return 0 - always_number(v)
```
Fixed:

```diff
--- a/test/fuzz/luaL_loadbuffer/preamble.lua
+++ b/test/fuzz/luaL_loadbuffer/preamble.lua
@@ -71,7 +71,7 @@ local __sub = function(v1, v2)
     return always_number(v1) - always_number(v2)
 end
 local __unm = function(v)
-    return 0 - always_number(v)
+    return - always_number(v)
 end
 
 debug.setmetatable('string', {

```

ligurio commented:
Changes are:

- splitted `test/fuzz: fix a group of semantic errors` to a number of commits
- updated commit messages
- fixed issues highlighted by comments on review
- added commit `cmake: fix protobuf library name`

ligurio commented:
Fixed preamble generation:

```diff
diff --git a/test/fuzz/luaL_loadbuffer/CMakeLists.txt b/test/fuzz/luaL_loadbuffer/CMakeLists.txt
index 0993d6f3fd..d33edb60d6 100644
--- a/test/fuzz/luaL_loadbuffer/CMakeLists.txt
+++ b/test/fuzz/luaL_loadbuffer/CMakeLists.txt
@@ -3,19 +3,18 @@ if (NOT EXISTS "${LUA_PREAMBLE_PATH}")
    message(FATAL "File '${LUA_PREAMBLE_PATH}' is not found.")
 endif ()
 
-# TODO: Add dependence on preamble.in.h and preamble.lua.
-include(txt2c)
-txt2c(${LUA_PREAMBLE_PATH} LUA_PREAMBLE)
+set(test_lua_sources)
+lua_source(test_lua_sources preamble.lua preamble_lua)
 
-configure_file(
-  ${CMAKE_CURRENT_SOURCE_DIR}/preamble.h.in
-  ${CMAKE_CURRENT_BINARY_DIR}/preamble.h
-  @ONLY
-)
+add_custom_target(generate_test_lua_sources
+    WORKING_DIRECTORY ${PROJECT_BINARY_DIR}/test/fuzz/luaL_loadbuffer/
+    DEPENDS ${test_lua_sources})
+set_property(DIRECTORY PROPERTY ADDITIONAL_MAKE_CLEAN_FILES ${test_lua_sources})
 
 add_executable(luaL_loadbuffer_fuzzer
                luaL_loadbuffer_fuzzer.cc
-               serializer.cc)
+               serializer.cc
+               ${CMAKE_CURRENT_BINARY_DIR}/preamble.lua.c)
 
 add_library(lua_grammar-proto)
 
diff --git a/test/fuzz/luaL_loadbuffer/serializer.cc b/test/fuzz/luaL_loadbuffer/serializer.cc
index 48f39f8381..8f950a6996 100644
--- a/test/fuzz/luaL_loadbuffer/serializer.cc
+++ b/test/fuzz/luaL_loadbuffer/serializer.cc
@@ -4,7 +4,6 @@
  * Copyright 2022, Tarantool AUTHORS, please see AUTHORS file.
  */
 #include "serializer.h"
-#include "preamble.h"
 
 #include <stack>
 #include <string>
@@ -13,6 +12,8 @@
 
 using namespace lua_grammar;
 
+extern char preamble_lua[];
+
 #define PROTO_TOSTRING(TYPE, VAR_NAME) \
 	std::string TYPE##ToString(const TYPE & (VAR_NAME))
 
@@ -1107,7 +1108,7 @@ MainBlockToString(const Block &block)
 	GetCounterIdProvider().clean();
 
 	std::string block_str = BlockToString(block);
-	std::string retval = LUA_PREAMBLE;
+	std::string retval = preamble_lua;
 
 	for (size_t i = 0; i < GetCounterIdProvider().count(); ++i) {
 		retval += GetCounterName(i);
```

ligurio commented on test/fuzz/luaL_loadbuffer/serializer.cc:
```diff
@@ -233,18 +234,38 @@ class Context {
+
+	bool vararg_is_possible()
+	{
+		return !returnable_stack_.empty() &&
```
Fixed:

```diff
--- a/test/fuzz/luaL_loadbuffer/serializer.cc
+++ b/test/fuzz/luaL_loadbuffer/serializer.cc
@@ -295,9 +295,10 @@ class Context {
 
        bool vararg_is_possible()
        {
-               return !returnable_stack_.empty() &&
-                      returnable_stack_.top() ==
-                               BlockType::kReturnableWithVararg;
+               return (returnable_stack_.empty() ||
+                      (!returnable_stack_.empty() &&
+                       returnable_stack_.top() ==
+                               BlockType::kReturnableWithVararg));
        }
 
 private:
```

Fuzzing test generates samples like this:

```lua
Name0, Name0 , Name0 , Name0  =  ...  ; 
```

and 

```lua
  counter_0 = 0                                                                    
  repeat                                                                           
  if counter_0 > 5 then break end                                                  
  counter_0 = counter_0 + 1                                                        
  until 'Name' ==  ... ;  
```

Buristan reviewed (approved):
**[PATCH 02/10] cmake: fix protobuf library name**

LGTM.

Buristan reviewed (approved):
**[PATCH 03/10] test/fuzz: fix "cannot use '...' outside a vararg function"**

Thanks for the fixes!
LGTM!

Buristan reviewed (commented):
**[PATCH 04/10] test/fuzz: fix a group of semantic errors**

Thanks for the fixes!
The patch is awesome! Really simple solution for our straggling with these errors! Nice!

LGTM, just minor nits below.

ligurio commented on test/fuzz/luaL_loadbuffer/CMakeLists.txt:
```diff
@@ -15,7 +29,9 @@ protobuf_generate(LANGUAGE cpp
                       ${PROTOBUF_LIBRARIES})
 
-target_include_directories(luaL_loadbuffer_fuzzer PUBLIC ${CMAKE_CURRENT_BINARY_DIR})
+target_include_directories(luaL_loadbuffer_fuzzer
```
Sure, it is not needed anymore.

ligurio replied:
It is need for `lua_grammar.pb.h`:

```
/home/sergeyb/sources/MRG/tarantool/test/fuzz/luaL_loadbuffer/luaL_loadbuffer_fuzzer.cc:13:10: fatal error: 'lua_grammar.pb.h' file not 
found                                                                                                                                   
   13 | #include "lua_grammar.pb.h"                                 
      |          ^~~~~~~~~~~~~~~~~~                                                                                                     
1 error generated.  
```

change with formatting was reverted:

```diff
--- a/test/fuzz/luaL_loadbuffer/CMakeLists.txt
+++ b/test/fuzz/luaL_loadbuffer/CMakeLists.txt
@@ -29,9 +29,7 @@ protobuf_generate(LANGUAGE cpp
 target_link_libraries(lua_grammar-proto
                       ${PROTOBUF_LIBRARIES})
 
-target_include_directories(luaL_loadbuffer_fuzzer
-                           PUBLIC
-                           ${CMAKE_CURRENT_BINARY_DIR})
+target_include_directories(luaL_loadbuffer_fuzzer PUBLIC ${CMAKE_CURRENT_BINARY_DIR})
 
 target_link_libraries(luaL_loadbuffer_fuzzer
                       PUBLIC
```

Buristan reviewed (approved):
**[PATCH 05/10] test/fuzz: fix errors due to commented out code**

Thanks for the patch!
LGTM!

ligurio commented on test/fuzz/luaL_loadbuffer/preamble.h.in:
```diff
@@ -0,0 +1 @@
+#define LUA_PREAMBLE @LUA_PREAMBLE@
```
Removed.

Buristan reviewed (approved):
**[PATCH 06/10] test/fuzz: fix errors "ambiguous syntax"**

Thanks for the patch!
LGTM!

Buristan reviewed (approved):
**[PATCH 07/10] test/fuzz: fix a group of errors related to `for`**

Thanks for the fixes!
LGTM!

Buristan commented on test/fuzz/luaL_loadbuffer/preamble.lua:
```diff
@@ -22,9 +22,27 @@ local __index = function(self, key)
+    elseif type(v1) == 'string' and type(v2) == 'string' then
+        return v1 < v2 -- Lexicographic comparison.
+    else
+        return always_number(v1) <= always_number(v2)
```
Nit: I suppose this should be `<` for consistency.

ligurio replied:
Fixed.

ligurio replied:
```diff
--- a/test/fuzz/luaL_loadbuffer/preamble.lua
+++ b/test/fuzz/luaL_loadbuffer/preamble.lua
@@ -44,7 +44,7 @@ local __lt = function(v1, v2)
     elseif type(v1) == 'string' and type(v2) == 'string' then
         return v1 < v2 -- Lexicographic comparison.
     else
-        return always_number(v1) <= always_number(v2)
+        return always_number(v1) < always_number(v2)
     end
 end
 local __mod = function(v1, v2)
```

Buristan reviewed (commented):
**[PATCH 08/10] test/fuzz: fix errors "attempt to compare"**
Thanks for the fixes!
LGTM, just a minor nit below.

Buristan reviewed (approved):
**[PATCH 09/10] test/fuzz: fix errors "'<name>' expected near 'and'"**

Thanks for the fixes!
LGTM!

Buristan commented on test/fuzz/luaL_loadbuffer/serializer.cc:
```diff
@@ -931,7 +943,11 @@ NESTED_PROTO_TOSTRING(IndexWithExpression, indexexpr, Variable)
-	indexexpr_str += "[" + ExpressionToString(indexexpr.exp()) + "]";
+	indexexpr_str += "[";
+	/* Prevent errors 'table index is nil' and 'table index is NaN'. */
+	indexexpr_str += AllowedIndexExpressionToString(indexexpr.exp());
```
I suppose that this part is excess.
For example, the following code is OK for Lua and LuaJIT:
```
lua -e '
local t = {}
local a = 1
t[a] = t[nil]
'
```
The table index NaN/nil is prohibited only for storing operations.

ligurio replied:
Reverted in `IndexWithExpression` and reduced a patch for this commit:

```diff
diff --git a/test/fuzz/luaL_loadbuffer/serializer.cc b/test/fuzz/luaL_loadbuffer/serializer.cc
index c6f480d1ef..b4500da9ca 100644
--- a/test/fuzz/luaL_loadbuffer/serializer.cc
+++ b/test/fuzz/luaL_loadbuffer/serializer.cc
@@ -54,9 +54,9 @@ const std::set<std::string> KReservedLuaKeywords {
 };
 
 const std::string kCounterNamePrefix = "counter_";
-const std::string kNotNaNAndNilWrapperName = "not_nan_and_nil";
 const std::string kNumberWrapperName = "always_number";
 const std::string kBinOpWrapperName = "only_numbers_cmp";
+const std::string kNotNaNAndNilWrapperName = "not_nan_and_nil";
 
 PROTO_TOSTRING(Block, block);
 PROTO_TOSTRING(Chunk, chunk);
@@ -943,11 +943,7 @@ NESTED_PROTO_TOSTRING(IndexWithExpression, indexexpr, Variable)
 {
 	std::string indexexpr_str = PrefixExpressionToString(
 		indexexpr.prefixexp());
-	indexexpr_str += "[";
-	/* Prevent errors 'table index is nil' and 'table index is NaN'. */
-	indexexpr_str += AllowedIndexExpressionToString(indexexpr.exp());
-	indexexpr_str += "]";
-
+	indexexpr_str += "[" + ExpressionToString(indexexpr.exp()) + "]";
 	return indexexpr_str;
 }
 
@@ -1107,11 +1103,9 @@ PROTO_TOSTRING(Field, field)
 
 NESTED_PROTO_TOSTRING(ExpressionAssignment, assignment, Field)
 {
-	std::string assignment_str;
-	assignment_str += "[";
 	/* Prevent error 'table index is nil' and 'table index is NaN'. */
-	assignment_str += AllowedIndexExpressionToString(assignment.key());
-	assignment_str += "]";
+	std::string assignment_str = "[ " +
+		AllowedIndexExpressionToString(assignment.key()) + " ]";
 	assignment_str += " = " + ExpressionToString(assignment.value());
 	return assignment_str;
 }
```

Buristan reviewed (commented):
**[PATCH 10/10] test/fuzz: fix errors "table index is {nil, NaN}"**

LGTM, with a minor comment below.

Buristan reviewed (approved):
Thanks for the fixes!
Patchset LGTM!
